simdev/module/github/gather_popular_repos_stage.py

```diff
@@ -0,0 +1,237 @@
+import logging
+from collections import Counter
+from multiprocessing import Pool
+from os import path
+
+from tqdm import tqdm
+
+from simdev.module.github.github_api_wrapper import GithubApiWrapper
+from simdev.util.pipeline import Pipeline, Stage
+
+
+class StargazersContext:
+    """
+    Context for storing info about repository's stargazers
+    """
+
+    def __init__(self, repo: str, page_limit: int, stargazers_save_directory: str):
+        self.repo = repo
+        self.page_limit = page_limit
+        self.stargazers_save_directory = stargazers_save_directory
+        self.stargazers: set[str] = set()
+
+    def recover(self):
+        """
+        Recover stargazer list from the save file (only matching page_limit)
+        :return: self if recovery went successful or None otherwise
+        """
+
+        def hook(recovered):
+            if recovered['repo'] == self.repo and recovered['page_limit'] == self.page_limit:
+                self.stargazers = set(recovered['stargazers'])
+                return self
+            return None
+
+        return Pipeline.load_context(hook=hook,
+                                     name=path.join(self.stargazers_save_directory, '{0}.json'.format(self.repo)))
+
+    def save(self):
+        """
+        Store stargazers list alongside page limit (and repo) in the save file
+        """
+        Pipeline.store_context(
+            {
+                'repo': self.repo,
+                'stargazers': list(self.stargazers),
+                'page_limit': self.page_limit
+            },
+            path.join(self.stargazers_save_directory, '{0}.json'.format(self.repo)))
+
+
+class StarredReposContext:
+    """
+    Context for storing starred repositories of a user
+    """
+
+    def __init__(self, user: str, page_limit: int, starred_repos_save_directory: str):
+        self.user = user
+        self.page_limit = page_limit
+        self.starred_repos_save_directory = starred_repos_save_directory
+        self.repos: set[str] = set()
+
+    def recover(self):
+        """
+        Recover repositories starred by specific user
```

Comment:
hook logic explanation required - here and above - that parameters of pipeline could be changed and so on